File: src/indexer.ts
import { promises as fs } from 'fs';
import * as path from 'path';
import { app } from 'electron';
import { exec } from 'child_process';
import { promisify } from 'util';
import type { IndexEntry, IndexStatus } from './types';
import type { FileTaskManager } from './fileTasks';
import { getDrives } from './utils';
import { ignoreError } from './shared';

const execAsync = promisify(exec);
type IndexEntryPayload = Partial<Omit<IndexEntry, 'modified'>> & {
  modified?: Date | number | string;
};

const EXCLUDE_SEGMENTS = new Set([
  'node_modules',
  '.git',
  '.cache',
  'cache',
  'caches',
  '.trash',
  'trash',
  '$recycle.bin',
  'system volume information',
  '.npm',
  '.docker',
  'appdata',
  'programdata',
  'windows',
  'program files',
  'program files (x86)',
  '$windows.~bt',
  '$windows.~ws',
  'recovery',
  'perflogs',
  'library',
  '$winreagent',
  'config.msi',
  'msocache',
  'intel',
  'nvidia',
  'amd',
]);

const EXCLUDE_FILES = new Set([
  'pagefile.sys',
  'hiberfil.sys',
  'swapfile.sys',
  'dumpstack.log.tmp',
  'dumpstack.log',
  '.ds_store',
  'thumbs.db',
  'desktop.ini',
  'ntuser.dat',
  'ntuser.dat.log',
  'ntuser.dat.log1',
  'ntuser.dat.log2',
]);

async function writeFileAtomic(targetPath: string, data: string): Promise<void> {
  const dir = path.dirname(targetPath);
  const base = path.basename(targetPath);
  const tmpPath = path.join(
    dir,
    `${base}.tmp-${process.pid}-${Date.now()}-${Math.random().toString(16).slice(2)}`
  );

  await fs.writeFile(tmpPath, data, 'utf-8');

  try {
    await fs.rename(tmpPath, targetPath);
  } catch (error) {
    const err = error as NodeJS.ErrnoException;
    if (err.code === 'EEXIST' || err.code === 'EPERM' || err.code === 'EACCES') {
      try {
        await fs.unlink(targetPath);
      } catch (error) {
        ignoreError(error);
      }
      try {
        await fs.rename(tmpPath, targetPath);
        return;
      } catch (error) {
        ignoreError(error);
      }
    }
    try {
      await fs.copyFile(tmpPath, targetPath);
    } finally {
      await fs.unlink(tmpPath).catch(() => {});
    }
  }
}

export class FileIndexer {
  private index: Map<string, IndexEntry> = new Map();
  private isIndexing: boolean = false;
  private indexedFiles: number = 0;
  private totalFiles: number = 0;
  private lastIndexTime: Date | null = null;
  private indexPath: string;
  private enabled: boolean = true;
  private abortController: AbortController | null = null;
  private initializationPromise: Promise<void> | null = null;
  private fileTasks: FileTaskManager | null = null;
  private buildOperationId: string | null = null;

  // prevent unbounded mem growth
  private static readonly MAX_INDEX_SIZE = 200000;

  constructor(fileTasks?: FileTaskManager) {
    const userDataPath = app.getPath('userData');
    this.indexPath = path.join(userDataPath, 'file-index.json');
    this.fileTasks = fileTasks ?? null;
  }

  // platform-specific important dirs
  private async getCommonLocations(): Promise<string[]> {
    const platform = process.platform;
    const homeDir = app.getPath('home');
    const locations: string[] = [];

    if (platform === 'win32') {
      locations.push(
        path.join(homeDir, 'Desktop'),
        path.join(homeDir, 'Documents'),
        path.join(homeDir, 'Downloads'),
        path.join(homeDir, 'Pictures'),
        path.join(homeDir, 'Music'),
        path.join(homeDir, 'Videos')
      );
      const drives = await getDrives();
      locations.push(...drives);
    } else if (platform === 'darwin') {
      locations.push(
        path.join(homeDir, 'Desktop'),
        path.join(homeDir, 'Documents'),
        path.join(homeDir, 'Downloads'),
        path.join(homeDir, 'Pictures'),
        path.join(homeDir, 'Music'),
        path.join(homeDir, 'Movies'),
        '/Applications',
        '/Users'
      );
      const drives = await getDrives();
      for (const drive of drives) {
        if (drive !== '/' && !locations.includes(drive)) {
          locations.push(drive);
        }
      }
    } else {
      locations.push(
        path.join(homeDir, 'Desktop'),
        path.join(homeDir, 'Documents'),
        path.join(homeDir, 'Downloads'),
        path.join(homeDir, 'Pictures'),
        path.join(homeDir, 'Music'),
        path.join(homeDir, 'Videos'),
        '/usr',
        '/opt',
        '/home'
      );
      const drives = await getDrives();
      for (const drive of drives) {
        if (drive !== '/' && !locations.includes(drive)) {
          locations.push(drive);
        }
      }
    }

    return locations;
  }
  private shouldExclude(filePath: string): boolean {
    const lastSep = Math.max(filePath.lastIndexOf('/'), filePath.lastIndexOf('\\'));
    const filename = (lastSep >= 0 ? filePath.slice(lastSep + 1) : filePath).toLowerCase();

    if (EXCLUDE_FILES.has(filename)) return true;

    const lowerPath = filePath.toLowerCase();
    for (const segment of EXCLUDE_SEGMENTS) {
      const idx = lowerPath.indexOf(segment);
      if (idx === -1) continue;
      const before = idx === 0 || lowerPath[idx - 1] === '/' || lowerPath[idx - 1] === '\\';
      const after =
        idx + segment.length === lowerPath.length ||
        lowerPath[idx + segment.length] === '/' ||
        lowerPath[idx + segment.length] === '\\';
      if (before && after) return true;
    }
    return false;
  }

  // recursive scan with batching
  private async scanDirectory(dirPath: string, signal?: AbortSignal): Promise<void> {
    if (signal?.aborted || this.shouldExclude(dirPath)) {
      return;
    }

    try {
      const entries = await fs.readdir(dirPath, { withFileTypes: true });

      const BATCH_SIZE = 50;
      const subdirs: string[] = [];

      for (let i = 0; i < entries.length; i += BATCH_SIZE) {
        if (signal?.aborted || this.index.size >= FileIndexer.MAX_INDEX_SIZE) {
          return;
        }

        const batch = entries.slice(i, i + BATCH_SIZE);
        const results = await Promise.allSettled(
          batch.map(async (entry) => {
            const fullPath = path.join(dirPath, entry.name);

            if (this.shouldExclude(fullPath)) {
              return null;
            }

            const stats = await fs.stat(fullPath);
            return {
              entry,
              fullPath,
              stats,
              isDirectory: entry.isDirectory(),
            };
          })
        );

        for (const result of results) {
          if (signal?.aborted || this.index.size >= FileIndexer.MAX_INDEX_SIZE) {
            return;
          }

          if (result.status === 'fulfilled' && result.value) {
            const { entry, fullPath, stats, isDirectory } = result.value;

            const indexEntry: IndexEntry = {
              name: entry.name,
              path: fullPath,
              isDirectory,
              isFile: entry.isFile(),
              size: stats.size,
              modified: stats.mtime,
            };

            this.index.set(fullPath, indexEntry);
            this.indexedFiles++;

            if (isDirectory && this.index.size < FileIndexer.MAX_INDEX_SIZE) {
              subdirs.push(fullPath);
            }
          }
        }
      }

      for (const subdir of subdirs) {
        if (signal?.aborted || this.index.size >= FileIndexer.MAX_INDEX_SIZE) {
          return;
        }
        await this.scanDirectory(subdir, signal);
      }
    } catch (error) {
      console.log(`[Indexer] Cannot access ${dirPath}:`, (error as Error).message);
    }
  }
  private async estimateTotalFiles(locations: string[]): Promise<number> {
    let estimate = 0;
    for (const location of locations) {
      try {
        await fs.access(location);
        estimate += 1000;
      } catch (error) {
        ignoreError(error);
      }
    }
    return estimate;
  }

  private normalizeIndexEntry(entryPath: string, entry: IndexEntryPayload): IndexEntry | null {
    if (!entryPath || typeof entryPath !== 'string') {
      return null;
    }
    const name = typeof entry.name === 'string' ? entry.name : path.basename(entryPath);
    const isDirectory = typeof entry.isDirectory === 'boolean' ? entry.isDirectory : false;
    const isFile = typeof entry.isFile === 'boolean' ? entry.isFile : !isDirectory;
    const size = typeof entry.size === 'number' ? entry.size : 0;
    const modifiedValue = entry.modified ?? 0;
    let modified =
      modifiedValue instanceof Date ? modifiedValue : new Date(modifiedValue as string | number);
    if (Number.isNaN(modified.getTime())) {
      modified = new Date(0);
    }

    return {
      name,
      path: entryPath,
      isDirectory,
      isFile,
      size,
      modified,
    };
  }

  private parseIndexTime(value: unknown): Date | null {
    if (value instanceof Date) {
      const time = value.getTime();
      return Number.isNaN(time) ? null : value;
    }
    if (typeof value !== 'number' && typeof value !== 'string') {
      return null;
    }
    const date = new Date(value);
    return Number.isNaN(date.getTime()) ? null : date;
  }

  private loadIndexEntries(entries: unknown[]): void {
    this.index.clear();
    for (const entry of entries) {
      let entryPath: string | null = null;
      let item: IndexEntryPayload | null = null;

      if (Array.isArray(entry)) {
        entryPath = typeof entry[0] === 'string' ? entry[0] : null;
        item = (entry[1] as IndexEntryPayload) || null;
      } else if (entry && typeof entry === 'object') {
        const obj = entry as IndexEntryPayload & { path?: string };
        entryPath = typeof obj.path === 'string' ? obj.path : null;
        item = obj;
      }

      if (!entryPath || !item) {
        continue;
      }

      const normalized = this.normalizeIndexEntry(entryPath, item);
      if (normalized) {
        this.index.set(entryPath, normalized);
      }
    }

    this.indexedFiles = this.index.size;
  }
  async buildIndex(): Promise<void> {
    if (this.isIndexing) {
      console.log('[Indexer] Already indexing, skipping...');
      return;
    }

    if (!this.enabled) {
      console.log('[Indexer] Indexer is disabled, skipping...');
      return;
    }

    this.isIndexing = true;
    this.indexedFiles = 0;
    this.index.clear();
    this.abortController = new AbortController();

    console.log('[Indexer] Starting index build...');

    try {
      const locations = await this.getCommonLocations();
      console.log(`[Indexer] Locations to scan: ${locations.join(', ')}`);
      this.totalFiles = await this.estimateTotalFiles(locations);

      if (this.fileTasks) {
        this.buildOperationId = `index-${Date.now()}-${Math.random().toString(36).slice(2)}`;
        const result = await this.fileTasks.runTask<{
          entries: Array<[string, IndexEntryPayload]>;
        }>(
          'build-index',
          {
            locations,
            skipDirs: Array.from(EXCLUDE_SEGMENTS),
            maxIndexSize: FileIndexer.MAX_INDEX_SIZE,
          },
          this.buildOperationId
        );
        const entries = Array.isArray(result.entries) ? result.entries : [];
        for (const [entryPath, entry] of entries) {
          if (this.index.size >= FileIndexer.MAX_INDEX_SIZE) break;
          const normalized = this.normalizeIndexEntry(entryPath, entry);
          if (normalized) {
            this.index.set(entryPath, normalized);
          }
        }
        this.indexedFiles = this.index.size;
        console.log(`[Indexer] Worker scan complete. Indexed ${this.indexedFiles} files.`);
      } else {
        for (const location of locations) {
          if (this.abortController.signal.aborted) {
            break;
          }

          const beforeCount = this.indexedFiles;
          try {
            await fs.access(location);
            console.log(`[Indexer] Scanning: ${location}`);
            await this.scanDirectory(location, this.abortController.signal);
            console.log(
              `[Indexer] Finished ${location}: indexed ${this.indexedFiles - beforeCount} new files (total: ${this.indexedFiles})`
            );
          } catch (error) {
            console.log(`[Indexer] Skipping ${location}:`, (error as Error).message);
          }
        }
      }

      this.lastIndexTime = new Date();
      await this.saveIndex();
      console.log(`[Indexer] Index build complete. Indexed ${this.indexedFiles} files.`);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      if (message === 'Cancelled' || message === 'Calculation cancelled') {
        this.index.clear();
        this.indexedFiles = 0;
        console.log('[Indexer] Index build cancelled');
      } else {
        console.error('[Indexer] Error building index:', error);
      }
    } finally {
      this.isIndexing = false;
      if (this.abortController) {
        this.abortController.abort();
        this.abortController = null;
      }
      this.buildOperationId = null;
    }
  }

  async search(query: string): Promise<IndexEntry[]> {
    if (!this.enabled) {
      return [];
    }

    if (this.initializationPromise) {
      await this.initializationPromise;
    }

    if (this.index.size === 0 && !this.isIndexing) {
      await this.loadIndex();
    }

    const lowerQuery = query.toLowerCase();
    const results: IndexEntry[] = [];

    for (const entry of this.index.values()) {
      if (entry.name.toLowerCase().includes(lowerQuery)) {
        results.push(entry);
        if (results.length >= 100) {
          break;
        }
      }
    }
    results.sort((a, b) => {
      const aExact = a.name.toLowerCase() === lowerQuery;
      const bExact = b.name.toLowerCase() === lowerQuery;

      if (aExact && !bExact) return -1;
      if (!aExact && bExact) return 1;

      return a.name.localeCompare(b.name);
    });

    return results;
  }

  async getEntries(): Promise<IndexEntry[]> {
    if (!this.enabled) {
      return [];
    }

    if (this.initializationPromise) {
      await this.initializationPromise;
    }

    if (this.index.size === 0 && !this.isIndexing) {
      await this.loadIndex();
    }

    return Array.from(this.index.values());
  }

  async saveIndex(): Promise<void> {
    try {
      const entries = Array.from(this.index.entries());
      const lastIndexTime = this.lastIndexTime ? this.lastIndexTime.getTime() : null;
      if (this.fileTasks) {
        await this.fileTasks.runTask(
          'save-index',
          { indexPath: this.indexPath, entries, lastIndexTime },
          `save-index-${Date.now()}`
        );
      } else {
        const data = {
          index: entries,
          lastIndexTime,
          version: 1,
        };

        await writeFileAtomic(this.indexPath, JSON.stringify(data));
      }
      console.log(`[Indexer] Index saved to ${this.indexPath}`);
    } catch (error) {
      console.error('[Indexer] Error saving index:', error);
    }
  }

  async loadIndex(): Promise<void> {
    try {
      if (this.fileTasks) {
        const result = await this.fileTasks.runTask<{
          exists: boolean;
          index?: Array<unknown>;
          lastIndexTime?: string | number | null;
        }>('load-index', { indexPath: this.indexPath }, `load-index-${Date.now()}`);
        if (!result.exists) {
          console.log('[Indexer] No existing index found, will build on first search');
          return;
        }
        this.loadIndexEntries(result.index || []);
        this.lastIndexTime = this.parseIndexTime(result.lastIndexTime);
      } else {
        const data = await fs.readFile(this.indexPath, 'utf-8');
        const parsed = JSON.parse(data);

        this.loadIndexEntries(Array.isArray(parsed.index) ? parsed.index : []);
        this.lastIndexTime = this.parseIndexTime(parsed.lastIndexTime);
      }
      console.log(`[Indexer] Index loaded: ${this.indexedFiles} files`);
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        console.log('[Indexer] No existing index found, will build on first search');
      } else {
        console.error('[Indexer] Error loading index:', error);
      }
    }
  }

  async clearIndex(): Promise<void> {
    this.index.clear();
    this.indexedFiles = 0;
    this.lastIndexTime = null;

    try {
      await fs.unlink(this.indexPath);
      console.log('[Indexer] Index file deleted');
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code !== 'ENOENT') {
        console.error('[Indexer] Error deleting index:', error);
      }
    }
  }

  async rebuildIndex(): Promise<void> {
    console.log('[Indexer] Rebuilding index...');

    if (this.abortController) {
      this.abortController.abort();
    }
    if (this.fileTasks && this.buildOperationId) {
      this.fileTasks.cancelOperation(this.buildOperationId);
      this.buildOperationId = null;
    }

    await this.clearIndex();
    await this.buildIndex();
  }

  getStatus(): IndexStatus {
    return {
      isIndexing: this.isIndexing,
      totalFiles: this.totalFiles,
      indexedFiles: this.indexedFiles,
      lastIndexTime: this.lastIndexTime,
    };
  }

  setEnabled(enabled: boolean): void {
    this.enabled = enabled;
    console.log(`[Indexer] Indexer ${enabled ? 'enabled' : 'disabled'}`);

    if (!enabled && this.abortController) {
      this.abortController.abort();
      this.abortController = null;
    }
    if (!enabled && this.fileTasks && this.buildOperationId) {
      this.fileTasks.cancelOperation(this.buildOperationId);
      this.buildOperationId = null;
    }
  }

  isEnabled(): boolean {
    return this.enabled;
  }

  async initialize(enabled: boolean): Promise<void> {
    this.enabled = enabled;

    if (!enabled) {
      console.log('[Indexer] Indexer disabled, skipping initialization');
      return;
    }

    console.log('[Indexer] Initializing...');

    this.initializationPromise = (async () => {
      try {
        await this.loadIndex();

        if (this.index.size === 0) {
          console.log('[Indexer] No existing index found, building now...');
          await this.buildIndex();
        } else if (
          !this.lastIndexTime ||
          Date.now() - this.lastIndexTime.getTime() > 7 * 24 * 60 * 60 * 1000
        ) {
          console.log('[Indexer] Index is outdated, rebuilding...');
          await this.buildIndex();
        } else {
          console.log('[Indexer] Using existing index with', this.index.size, 'files');
        }
      } catch (err) {
        console.error('[Indexer] Initialization failed:', err);
      } finally {
        this.initializationPromise = null;
      }
    })();
  }
}

File: src-tauri/src/indexer.rs
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex, RwLock};
use std::time::UNIX_EPOCH;
use tauri::Manager;
use walkdir::WalkDir;

const MAX_INDEX_SIZE: usize = 50_000;
const MAX_SCAN_DEPTH: usize = 20;
// Yield CPU every N entries so bg build can't peg a core.
const SCAN_YIELD_EVERY: usize = 256;
const SCAN_YIELD_MS: u64 = 2;

static FILE_INDEX: std::sync::LazyLock<Arc<RwLock<FileIndex>>> =
    std::sync::LazyLock::new(|| Arc::new(RwLock::new(FileIndex::default())));

static BUILD_CANCEL: AtomicBool = AtomicBool::new(false);
static INDEXER_ENABLED: AtomicBool = AtomicBool::new(true);

static BUILD_MUTEX: std::sync::LazyLock<Mutex<()>> = std::sync::LazyLock::new(|| Mutex::new(()));

#[derive(Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct IndexEntry {
    pub name: String,
    // Lowercased name for search; not persisted, repopulated on load/build.
    #[serde(skip)]
    pub name_lower: String,
    pub path: String,
    pub is_directory: bool,
    pub is_file: bool,
    pub size: u64,
    pub modified: f64,
}

#[derive(Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
struct IndexFileData {
    #[serde(default)]
    index: Vec<IndexEntry>,
    #[serde(default)]
    last_index_time: Option<f64>,
    #[serde(default)]
    version: u32,
}

#[derive(Default)]
pub struct FileIndex {
    pub entries: Vec<IndexEntry>,
    pub is_building: bool,
    pub last_built: Option<String>,
}

fn exclude_segments() -> HashSet<&'static str> {
    [
        "node_modules",
        ".git",
        ".cache",
        ".trash",
        "trash",
        "$recycle.bin",
        "system volume information",
        ".npm",
        ".docker",
        ".snap",
        "appdata",
        "programdata",
        "windows",
        "program files",
        "program files (x86)",
        "$windows.~bt",
        "$windows.~ws",
        "recovery",
        "perflogs",
        "$winreagent",
        "config.msi",
        "msocache",
        "intel",
        "nvidia",
        "amd",
        "lost+found",
        ".spotlight-v100",
        ".trashes",
        ".fseventsd",
        ".documentrevisions-v100",
    ]
    .into_iter()
    .collect()
}

fn exclude_files() -> HashSet<&'static str> {
    [
        "pagefile.sys",
        "hiberfil.sys",
        "swapfile.sys",
        "dumpstack.log.tmp",
        "dumpstack.log",
        ".ds_store",
        "thumbs.db",
        "desktop.ini",
        "ntuser.dat",
        "ntuser.dat.log",
        "ntuser.dat.log1",
        "ntuser.dat.log2",
    ]
    .into_iter()
    .collect()
}

fn should_exclude(path: &Path, excl_segments: &HashSet<&str>, excl_files: &HashSet<&str>) -> bool {
    if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
        if excl_files.contains(name.to_lowercase().as_str()) {
            return true;
        }
    }
    for component in path.components() {
        if let std::path::Component::Normal(seg) = component {
            if let Some(s) = seg.to_str() {
                if excl_segments.contains(s.to_lowercase().as_str()) {
                    return true;
                }
            }
        }
    }
    false
}

fn get_index_locations() -> Vec<PathBuf> {
    let mut locations = Vec::new();

    if let Some(user_dirs) = directories::UserDirs::new() {
        let home = user_dirs.home_dir();
        for dir in &[
            "Desktop",
            "Documents",
            "Downloads",
            "Pictures",
            "Music",
            "Videos",
        ] {
            let p = home.join(dir);
            if p.exists() {
                locations.push(p);
            }
        }

        #[cfg(target_os = "macos")]
        {
            let movies = home.join("Movies");
            if movies.exists() {
                locations.push(movies);
            }
            // Skip /Applications + /Volumes: deep .app trees, external mounts. Opt-in later.
        }

        #[cfg(target_os = "linux")]
        {
            for p in &["/usr", "/opt"] {
                let path = Path::new(p);
                if path.exists() {
                    locations.push(path.to_path_buf());
                }
            }
        }
    }

    #[cfg(target_os = "windows")]
    {
        for letter in b'C'..=b'Z' {
            let drive = format!("{}:\\", letter as char);
            if Path::new(&drive).exists() {
                locations.push(PathBuf::from(drive));
            }
        }
    }

    locations
}

fn name_lower_of(name: &str) -> String {
    name.to_lowercase()
}

fn populate_name_lower(entries: &mut [IndexEntry]) {
    for entry in entries.iter_mut() {
        if entry.name_lower.is_empty() {
            entry.name_lower = name_lower_of(&entry.name);
        }
    }
}

fn build_index_sync() -> Vec<IndexEntry> {
    let excl_segments = exclude_segments();
    let excl_files = exclude_files();
    let locations = get_index_locations();
    let mut entries = Vec::new();
    let mut scanned = 0usize;

    for location in &locations {
        if is_build_cancelled() {
            break;
        }

        for entry in WalkDir::new(location)
            .max_depth(MAX_SCAN_DEPTH)
            .follow_links(false)
            .into_iter()
            .filter_entry(|e| !should_exclude(e.path(), &excl_segments, &excl_files))
            .filter_map(|e| {
                e.map_err(|err| log::warn!("[Indexer] walk error: {}", err))
                    .ok()
            })
        {
            if is_build_cancelled() || entries.len() >= MAX_INDEX_SIZE {
                break;
            }

            // Throttle: sleep every SCAN_YIELD_EVERY entries; re-check cancel.
            scanned += 1;
            if scanned.is_multiple_of(SCAN_YIELD_EVERY) {
                if is_build_cancelled() {
                    break;
                }
                std::thread::sleep(std::time::Duration::from_millis(SCAN_YIELD_MS));
            }

            let path = entry.path();
            let meta = match entry.metadata() {
                Ok(m) => m,
                Err(_) => continue,
            };

            let name = path
                .file_name()
                .map(|n| n.to_string_lossy().to_string())
                .unwrap_or_default();

            let modified = meta
                .modified()
                .ok()
                .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
                .map(|d| d.as_secs_f64() * 1000.0)
                .unwrap_or(0.0);

            let name_lower = name_lower_of(&name);
            entries.push(IndexEntry {
                name,
                name_lower,
                path: path.to_string_lossy().to_string(),
                is_directory: meta.is_dir(),
                is_file: meta.is_file(),
                size: meta.len(),
                modified,
            });
        }

        if entries.len() >= MAX_INDEX_SIZE {
            break;
        }
    }

    entries
}

fn is_build_cancelled() -> bool {
    BUILD_CANCEL.load(Ordering::Relaxed)
}

fn set_build_cancelled(val: bool) {
    BUILD_CANCEL.store(val, Ordering::Relaxed);
}

fn index_path(app: &tauri::AppHandle) -> Result<PathBuf, String> {
    let dir = app
        .path()
        .app_data_dir()
        .map_err(|e: tauri::Error| e.to_string())?;
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    Ok(dir.join("file-index.json"))
}

fn load_index_from_disk(path: &Path) -> Option<IndexFileData> {
    let data = fs::read_to_string(path).ok()?;
    serde_json::from_str(&data).ok()
}

fn save_index_to_disk(path: &Path, entries: &[IndexEntry]) -> Result<(), String> {
    #[derive(serde::Serialize)]
    struct IndexFileDataRef<'a> {
        index: &'a [IndexEntry],
        last_index_time: Option<f64>,
        version: u32,
    }

    let data = IndexFileDataRef {
        index: entries,
        last_index_time: Some(
            std::time::SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap_or_default()
                .as_secs_f64()
                * 1000.0,
        ),
        version: 1,
    };

    let json = serde_json::to_string(&data).map_err(|e| e.to_string())?;
    let tmp = crate::make_temp_path(path, "index");

    // L2: create_new + O_NOFOLLOW (on Unix). Without this, if a local
    // attacker symlinked the temp path before us, File::create would
    // truncate-through-the-symlink and write our index data to a path
    // they chose. create_new makes the create fail if the path exists at all.
    let mut file = {
        #[cfg(unix)]
        {
            use std::os::unix::fs::OpenOptionsExt;
            // libc::O_NOFOLLOW = 0o400000 on Linux, 0x0100 on macOS — use the
            // libc constant via nix if available, but the stdlib OpenOptions
            // already has create_new() which is enough to defeat the pre-planted
            // symlink (open fails with EEXIST). Add O_NOFOLLOW for belt+suspenders
            // in case the parent path traversal contains a symlink we don't expect.
            const O_NOFOLLOW: i32 = if cfg!(target_os = "linux") {
                0o400000
            } else {
                0x0100
            };
            fs::OpenOptions::new()
                .write(true)
                .create_new(true)
                .custom_flags(O_NOFOLLOW)
                .open(&tmp)
                .map_err(|e| e.to_string())?
        }
        #[cfg(not(unix))]
        {
            fs::OpenOptions::new()
                .write(true)
                .create_new(true)
                .open(&tmp)
                .map_err(|e| e.to_string())?
        }
    };
    file.write_all(json.as_bytes()).map_err(|e| e.to_string())?;
    file.sync_all().map_err(|e| e.to_string())?;
    drop(file);

    if let Err(e) = fs::rename(&tmp, path) {
        if let Err(cleanup_err) = fs::remove_file(&tmp) {
            log::warn!(
                "[Indexer] Failed to clean up temp file {}: {}",
                tmp.display(),
                cleanup_err
            );
        }
        return Err(e.to_string());
    }

    Ok(())
}

pub fn initialize_index(app: &tauri::AppHandle) {
    let idx_path = match index_path(app) {
        Ok(p) => p,
        Err(_) => return,
    };

    if let Some(data) = load_index_from_disk(&idx_path) {
        if let Ok(mut index) = FILE_INDEX.write() {
            index.entries = data.index;
            populate_name_lower(&mut index.entries);
            index.last_built = data.last_index_time.map(|t| {
                chrono::DateTime::from_timestamp_millis(t as i64)
                    .map(|dt| dt.to_rfc3339())
                    .unwrap_or_default()
            });
            log::info!("Loaded {} index entries from disk", index.entries.len());
        }
    }

    // No auto-build: post-launch fs walk looks like idle CPU spike.
    // Built lazily on first search (ensure_index_built) or explicit rebuild.
}

/// Build once in bg if enabled but empty. No-op if populated or building.
pub fn ensure_index_built(app: &tauri::AppHandle) {
    if !INDEXER_ENABLED.load(Ordering::Relaxed) {
        return;
    }
    let (empty, building) = match FILE_INDEX.read() {
        Ok(idx) => (idx.entries.is_empty(), idx.is_building),
        Err(_) => return,
    };
    if !empty || building {
        return;
    }
    let app_handle = app.clone();
    std::thread::spawn(move || {
        run_index_build(&app_handle);
    });
}

fn run_index_build(app: &tauri::AppHandle) {
    let _guard = match BUILD_MUTEX.lock() {
        Ok(g) => g,
        Err(_) => return,
    };

    set_build_cancelled(false);

    if let Ok(mut index) = FILE_INDEX.write() {
        index.is_building = true;
    }

    log::info!("Starting file index build...");
    let entries = build_index_sync();
    let count = entries.len();

    if let Ok(mut index) = FILE_INDEX.write() {
        index.entries = entries;
        index.is_building = false;
        index.last_built = Some(chrono::Utc::now().to_rfc3339());
    }

    if let Ok(idx_path) = index_path(app) {
        if let Ok(index) = FILE_INDEX.read() {
            if let Err(e) = save_index_to_disk(&idx_path, &index.entries) {
                log::warn!("Failed to save index to disk: {}", e);
            }
        }
    }

    log::info!("Index build complete: {} entries", count);
}

pub fn search_in_index(query: &str) -> Vec<IndexEntry> {
    if !INDEXER_ENABLED.load(Ordering::Relaxed) {
        return Vec::new();
    }
    let query_lower = query.to_lowercase();
    let index = match FILE_INDEX.read() {
        Ok(idx) => idx,
        Err(_) => return Vec::new(),
    };

    index
        .entries
        .iter()
        .filter(|e| e.name_lower.contains(&query_lower))
        .take(10_000)
        .cloned()
        .collect()
}

pub fn get_status() -> (bool, usize, Option<String>) {
    match FILE_INDEX.read() {
        Ok(idx) => (idx.is_building, idx.entries.len(), idx.last_built.clone()),
        Err(_) => (false, 0, None),
    }
}

pub async fn trigger_rebuild(app: &tauri::AppHandle) {
    set_build_cancelled(true);
    tokio::time::sleep(std::time::Duration::from_millis(100)).await;

    let app_handle = app.clone();
    std::thread::spawn(move || {
        run_index_build(&app_handle);
    });
}

pub fn cancel_build() {
    set_build_cancelled(true);
}

pub fn set_enabled(enabled: bool, app: Option<&tauri::AppHandle>) {
    INDEXER_ENABLED.store(enabled, Ordering::Relaxed);
    if enabled {
        return;
    }
    set_build_cancelled(true);
    if let Ok(mut index) = FILE_INDEX.write() {
        index.entries.clear();
        index.is_building = false;
        index.last_built = None;
    }
    if let Some(app) = app {
        if let Ok(path) = index_path(app) {
            let _ = fs::remove_file(path);
        }
    }
}

pub fn is_enabled() -> bool {
    INDEXER_ENABLED.load(Ordering::Relaxed)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    #[test]
    fn should_exclude_skips_node_modules() {
        let segs = exclude_segments();
        let files = exclude_files();
        let p = Path::new("/home/u/proj/node_modules/pkg/index.js");
        assert!(should_exclude(p, &segs, &files));
    }

    #[test]
    fn should_not_exclude_plain_cache_folder_name() {
        let segs = exclude_segments();
        let files = exclude_files();
        let p = Path::new("/home/u/proj/build/cache/output.bin");
        assert!(!should_exclude(p, &segs, &files));
    }

    #[test]
    fn should_exclude_dot_cache() {
        let segs = exclude_segments();
        let files = exclude_files();
        let p = Path::new("/home/u/.cache/foo");
        assert!(should_exclude(p, &segs, &files));
    }

    #[test]
    fn should_not_exclude_macos_library_tree() {
        let segs = exclude_segments();
        let files = exclude_files();
        let p = Path::new("/Users/dev/Library/Fonts/Arial.ttf");
        assert!(!should_exclude(p, &segs, &files));
    }

    #[test]
    fn should_exclude_dot_snap() {
        let segs = exclude_segments();
        let files = exclude_files();
        let p = Path::new("/home/u/.snap/somepkg");
        assert!(should_exclude(p, &segs, &files));
    }
}
